File: Tree/687.py
class Solution:
    def longestUnivaluePath(self, root: Optional[TreeNode]) -> int:
        self.ans = 0

        def dfs(node):
            if not node: return

            left = dfs(node.left)
            right = dfs(node.right)

            left = left + 1 if node.left and node.val == node.left.val else 0
            right = right + 1 if node.right and node.val == node.right.val else 0
            self.ans = max(self.ans, left + right)

            return max(left, right)

        dfs(root)

        return self.ans

File: Tree/687.cpp
class Solution {
public:
    int longestUnivaluePath(TreeNode* root) {
        int ans = 0;
        dfs(root, ans);
        
        return ans;
        
    }
    int dfs(TreeNode* root, int& ans) {
        if (!root) return 0;
        int left_children = root->left ? dfs(root->left, ans) : 0;
        int right_children = root->right ? dfs(root->right, ans) : 0;
        int left = root->left && root->left->val == root->val ?left_children + 1 : 0;
        int right = root->right && root->right->val == root->val ? right_children + 1 : 0;
        ans = max(ans, left+right);
        return max(left, right);

    }
};
